### PR TITLE
tests: e2e eip 3074 tests

### LINKED ISSUE #1125: tests: add e2e auth / authcall tests
Will need to test the e2e auth and authcall tests. Can take inspiration from https://github.com/paradigmxyz/alphanet/tree/main/crates/testing/resources/eip3074

### LINKED ISSUE #1127: feat: if `authcall` gas is 0, use `gas_left` for the call
### Describe the Feature Request

As per eip 3074 if the gas param is 0, the gas_limit used in the call should be gas_left.

### LINKED ISSUE #1128: dev: use patched foundry in CI and makefile
We will need to use a patched version of Foundry with the solc compiler patched with auth / authcall support.

We should
 - Modify the makefile `build-sol` to use a dockerized version of foundry to run forge build
 - Modify the CI to use this dockerized image when building the sol contracts as well

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR:

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #1127 Resolves #1125 Closes #1128

## What is the new behavior?

-
-
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1129)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -33,3 +33,5 @@ logs
 
 tests/ef_tests/test_data
 !tests/ef_tests/test_data/.gitkeep
+
+bin/
```
Comment (ClementWalter) [inline .gitignore:37]: new line (vs code config is

"files.insertFinalNewline": true,
  "files.trimFinalNewlines": true,
  "files.trimTrailingWhitespace": true,

File: solidity_contracts/src/EIP3074/BaseAuth.sol
Comment (ClementWalter) [inline solidity_contracts/src/EIP3074/BaseAuth.sol:1]: we probably need to add solint

File: solidity_contracts/src/EIP3074/GasSponsorInvoker.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+        bytes calldata data
+    ) external pure returns (bool success) {
+        // Ensure the transaction is authorized by the signer
+        require(authSimple(authority, commit, v, r, s), "Authorization failed");
+
+        // Execute the call as authorized by the signer
+        success = authCallSimple(to, data, 0, 0);
+        require(success, "Call execution failed");
+    }
+}
```
Comment (ClementWalter) [inline solidity_contracts/src/EIP3074/GasSponsorInvoker.sol:35]: missing new line

File: solidity_contracts/src/EIP3074/SenderRecorder.sol [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+pragma solidity ^0.8.23;
+
+contract SenderRecorder {
+    address public lastSender;
+
+    function recordSender() external {
+        lastSender = msg.sender;
+    }
+}
```
Comment (ClementWalter) [inline solidity_contracts/src/EIP3074/SenderRecorder.sol:11]: missing new line

File: src/kakarot/instructions/system_operations.cairo [hunk 4]
```diff
@@ -1035,7 +1035,15 @@ namespace SystemOperations {
...
 
-        let gas = Gas.compute_message_call_gas(gas_param, evm.gas_left);
+        let (is_gas_zero) = uint256_eq(gas_param, Uint256(0, 0));
+        if (is_gas_zero != FALSE) {
+            let (high, low) = split_felt(evm.gas_left);
+            tempvar gas_limit = Uint256(low, high);
+        } else {
+            tempvar gas_limit = gas_param;
+        }
+        tempvar gas_limit = gas_limit;
```
Comment (ClementWalter) [inline src/kakarot/instructions/system_operations.cairo:1045]: `let gas_limit = [ap - 1]`

File: solidity_contracts/src/EIP3074/BaseAuth.sol
Comment (enitrat) [reply solidity_contracts/src/EIP3074/BaseAuth.sol:1]: to trunk?

File: src/kakarot/instructions/system_operations.cairo (hunk 4 above)
Comment (enitrat) [reply src/kakarot/instructions/system_operations.cairo]: let gas_limit = Uint256([ap-2], [ap-1]) ?

**Review by ClementWalter — APPROVED** [review]

File: src/kakarot/instructions/system_operations.cairo (hunk 4 above)
Comment (ClementWalter) [reply src/kakarot/instructions/system_operations.cairo]: I think that `cast([ap - 2], Uint256)` works as well

**Pull request merged**